FileShareManager.cs

```diff
@@ -0,0 +1,149 @@
+﻿using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.File;
+using NLog.Common;
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NLog.Extensions.AzureFileSystem
+{
+    public class FileShareManager
+    {
+        private readonly ReaderWriterLockSlim _fileLock = new ReaderWriterLockSlim();
+        private readonly IAzureFileSystemTarget target;
+        private CloudFileShare _fileShare;
+
+        public FileShareManager(IAzureFileSystemTarget target)
+        {
+            this.target = target;
+        }
+
+        private async Task<CloudFileDirectory> InitializeFolderAsync(string folderName)
+        {
+            if (_fileShare == null)
+            {
+                _fileShare = await CreateStorageAccountFromConnectionStringAsync(target)
+                    .ConfigureAwait(false);
+            }
+
+            folderName = folderName.Replace(@"\", @"/");
+
+            var rootDir = _fileShare.GetRootDirectoryReference();
+
+            // Get a reference to the directory.
+            var folder = string.IsNullOrWhiteSpace(folderName) ? rootDir : rootDir.GetDirectoryReference(folderName);
+
+            if (!await folder.ExistsAsync().ConfigureAwait(false))
+            {
+                var levels = folderName.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+                var currentLevel = levels[0];
+
+                for (int i = 0; i < levels.Length; i++)
+                {
+                    if (i > 0) currentLevel += @"/" + levels[i];
+
+                    var currentFolder = rootDir.GetDirectoryReference(currentLevel);
+                    if (!await currentFolder.ExistsAsync().ConfigureAwait(false))
+                    {
+                        await currentFolder.CreateAsync().ConfigureAwait(false);
+                        InternalLogger.Trace($"AzureFileSystemTarget - Folder {currentLevel} Initialized");
+                    }
+                }
+            }
+
+            return folder;
+        }
+        
+        /// <summary>
+        ///     Validates the connection string information in app.config and throws an exception if it looks like
+        ///     the user hasn't updated this to valid values.
+        /// </summary>
+        /// <param name="storageConnectionString">The storage connection string</param>
+        /// <returns>CloudStorageAccount object</returns>
+        private async Task<CloudFileShare> CreateStorageAccountFromConnectionStringAsync(IAzureFileSystemTarget target)
```

Comment:
method can be made static

Reply:
I noticed that this should be parameterless; no point in passing member object. fixed method code